src/appimagetool_fetch_runtime.c

```diff
@@ -0,0 +1,155 @@
+// need to define this to enable asprintf
+#define _GNU_SOURCE
+#include <stdbool.h>
+#include <stdio.h>
+
+#include <curl/curl.h>
+#include <malloc.h>
+#include <string.h>
+#include <errno.h>
+
+#include "appimagetool_fetch_runtime.h"
+
+bool fetch_runtime(char* arch, size_t* size, char** buffer, bool verbose) {
+    // not the cleanest approach to globally init curl here, but this method shouldn't be called more than once anyway
+    curl_global_init(CURL_GLOBAL_ALL);
+
+    char* url = NULL;
+    int url_size = asprintf(&url, "https://github.com/AppImage/type2-runtime/releases/download/continuous/runtime-%s", arch);
+    if (url_size <= 0) {
+        fprintf(stderr, "Failed to generate runtime URL\n");
+        curl_global_cleanup();
+        return false;
+    }
+
+    fprintf(stderr, "Downloading runtime file from %s\n", url);
+
+    char curl_error_buf[CURL_ERROR_SIZE];
+    CURL* handle = NULL;
+    char* effective_url;
+    int success = -1L;
+
+    curl_off_t content_length = -1;
+
+    // first, we perform a HEAD request to determine the required buffer size to write the file to
+    // of course, this assumes that a) GitHub sends a Content-Length header and b) that it is correct and will be in
+    // the GET request, too
+    // we store the URL we are redirected to (which probably lies on some AWS and is unique to that file) for use in
+    // the GET request, which should ensure we really download the file whose size we check now
+    handle = curl_easy_init();
+    
+    if (handle == NULL) {
+        fprintf(stderr, "Failed to initialize libcurl\n");
+        curl_global_cleanup();
+        return false;
+    }
+    
+    curl_easy_setopt(handle, CURLOPT_URL, url);
```

Comment:
Shouldn't we check the return values of `curl_easy_setopt` calls?

Reply:
We could (and maybe should), but I don't expect any issues with this code. Even their docs suggest to use these options like this. We could improve on this later. Also, we have a perspective on replacing all of this with a more user friendly C++ interface to libcurl.